nixos/rust-motd: allow ordering sections

Ma27 opened the pull request:
## Description of changes

cc reporter @Niols 
cc @infinisil because I'm curious if there's a module-system feature I'm not aware of to achieve this.

------

Closes #234802

The problem here is that with e.g.

    {
      uptime.prefix = "Up";
      banner.command = "hostname | figlet -f slant";
    }

`banner` still appears before `uptime` in the final motd text because
Nix sorts attribute names alphabetically internally.

To work around this without breaking compatibility or losing the
property to override individual sections in other modules - e.g.

    {
      banner.color = mkForce "blue";
    }

I decided to introduce an option `priority` here, similar to the
priority field for `nginx`[1] and with the same semantics (i.e. higher
value means lower priority).

Internally a bunch of env vars are generated, i.e. `env0` to `envN` for
`N` sections with each of them containing a declaration for the TOML,
i.e. `env0` contains `{ uptime.prefix = "Up"; }` and `env1` contains
`{ banner.command = "hostname | figlet -f slant"; }` if `uptime.priority`
is set to a value below 1000.

In this order, the declarations are concatenated together by `jq(1)`
which doesn't sort keys alphabetically which results in a JSON
representation with `uptime` before `banner`. This is finally piped to
`json2toml` which converts this into TOML for rust-motd.

[1] https://nixos.org/manual/nixos/unstable/options#opt-services.nginx.virtualHosts._name_.locations._name_.priority

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Ma27 committed:
nixos/rust-motd: run once on bootup

That way e.g. the last login and uptime isn't completely bogus when
accessing a machine for the first time after a reboot.

Ma27 committed:
nixos/rust-motd: allow ordering sections by `priority`

Closes #234802

The problem here is that with e.g.

    {
      uptime.prefix = "Up";
      banner.command = "hostname | figlet -f slant";
    }

`banner` still appears before `uptime` in the final motd text because
Nix sorts attribute names alphabetically internally.

To work around this without breaking compatibility or losing the
property to override individual sections in other modules - e.g.

    {
      banner.color = mkForce "blue";
    }

I decided to introduce an option `priority` here, similar to the
priority field for `nginx`[1] and with the same semantics (i.e. higher
value means lower priority).

Internally a bunch of env vars are generated, i.e. `env0` to `envN` for
`N` sections with each of them containing a declaration for the TOML,
i.e. `env0` contains `{ uptime.prefix = "Up"; }` and `env1` contains
`{ banner.command = "hostname | figlet -f slant"; }` if `uptime.priority`
is set to a value below 1000.

In this order, the declarations are concatenated together by `jq(1)`
which doesn't sort keys alphabetically which results in a JSON
representation with `uptime` before `banner`. This is finally piped to
`json2toml` which converts this into TOML for rust-motd.

[1] https://nixos.org/manual/nixos/unstable/options#opt-services.nginx.virtualHosts._name_.locations._name_.priority

Ma27 committed:
nixos/rust-motd: .attrs.json -> "$NIX_ATTRS_JSON_FILE"

That way the derivation can also be built in a `nix-shell` where
`.attrs.json` isn't under "$NIX_BUILD_TOP".

Niols commented on nixos/modules/programs/rust-motd.nix:
```diff
@@ -28,9 +55,35 @@ in {
+          type = types.int;
+          default = modules.defaultOrderPriority;
+          description = mdDoc ''
+            In `rust-motd`, the order of the sections in TOML correlates to the order
+            of the items displayed in the resulting `motd`. Since attributes in Nix are
+            ordered alphabetically, e.g. `banner` would always be before `uptime`.
```
The message is clear. However, the “Since” made me expect another part of the sentence.

Ma27 replied:
Very good point. I guess I was getting tired when I wrote the explanations here, will fix that :)

Niols commented on nixos/modules/programs/rust-motd.nix:
```diff
@@ -50,9 +103,10 @@ in {
       path = with pkgs; [ bash ];
       documentation = [ "https://github.com/rust-motd/rust-motd/blob/v${pkgs.rust-motd.version}/README.md" ];
       description = "motd generator";
+      wantedBy = [ "multi-user.target" ];
```
I suppose this isn't related to the rest of the change?

Ma27 replied:
Yes, correct. This was more of a "why didn't I do it already that way"-moment and that's also the reason why this is in its own commit. If everybody agrees on that, I'd say we'll merge it right in, otherwise I'm happy to remove it from this branch and open up another PR for it.

Niols replied:
I think it's fine having it in this PR. As you say, it lives in its own commit anyways.

Niols reviewed (commented):
This sounds like a sensible solution to me. I like that you use a mechanism (the priorities) that is already widely available. Having to pass through `jq` feels like a bit of a hack but I cannot think of a better way. (The fact that `rust-motd` relies on ordering in a TOML file feels already pretty hackish to me.) The Nix implementation looks sound and the change is surprisingly tiny. Thanks for this!

Ma27 commented:
Btw, in case anybody wants to test this on 23.05, you'll also need https://github.com/Ma27/nixpkgs/commit/058e5fcb2414916eeb1536e2f0864405909801d7 (explanation is in the commit's body).

> Having to pass through jq feels like a bit of a hack but I cannot think of a better way.

Since Nix orders attributes internally and I couldn't find a reasonable way to do this in pure Nix (admittedly I was considering to write my own TOML renderer, but after looking at the spec I decided that this is the simpler approach).

One could also implement a DAG-like type as it's done in home-manager already, but a change to the module-system itself would probably take even more time to be ready (and there's pending prior work already, e.g. https://github.com/NixOS/nixpkgs/pull/97392).

Niols commented:
> I decided that this is the simpler approach

> a change to the module-system itself would probably take even more time to be ready

Yes, definitely. What's in this PR is the best option in my opinion. Let see if someone else has a better idea!

infinisil commented on nixos/modules/programs/rust-motd.nix:
```diff
@@ -28,9 +55,35 @@ in {
+          type = types.int;
+          default = modules.defaultOrderPriority;
+          description = mdDoc ''
+            In `rust-motd`, the order of the sections in TOML correlates to the order
+            of the items displayed in the resulting `motd`. Attributes in Nix are
+            ordered alphabetically, e.g. `banner` would always be before `uptime`.
```
Oof, I looked into this, apparently it's not really specified in toml whether the order should be preserved. This here from [the table spec](https://toml.io/en/v1.0.0#table) is the only mention regarding the order, but it's not exactly that:

> Key/value pairs within tables are not guaranteed to be in any specific order.

In any case, the `toml` crate, which is used by rust-motd doesn't preserve the order by default, _but_ you [can enable it](https://docs.rs/toml/latest/toml/type.Table.html) to be preserved with a feature flag, which rust-motd indeed enables.

The origin of this flag can be tracked back to [this issue](https://github.com/toml-rs/toml-rs/issues/232) requesting the feature. In the discussion one of the use cases (and indeed one that makes sense) is to preserve the ordering when programmatically editing toml files.

However, for the purpose of being able to change behavior based on the order, that's just not great imo and should never have been allowed, because it leads to configurations that can't be nicely composed and is most likely not what the toml spec intended. This is really ugly here now because of that, when there's better ways of defining orderings, the best of which is probably just a list of elements like

```toml
order = [
  "banner",
  "uptime",
]
```

Anyways, it would be great to report this to the toml spec, rust-motd, the toml crate, etc., asking them to consider not relying on the order of keys for determining behavior. 

---

And as for a Nix-specific workaround, I think having such an `order` option of sorts would give a cleaner interface:
```nix
programs.rust-motd = {
  settings.banner = ...;
  settings.uptime = ...;
  order = [
    "banner"
    "uptime"
  ];
}
```

Ma27 replied:
Fair, that's indeed clearer, updated the code accordingly.

Ma27 replied:
> Key/value pairs within tables are not guaranteed to be in any specific order.

Hmm, reading the standard now, this seems to refer to the keys inside a table, i.e.

```toml
[foo]
bar = 1
baz = 2
```

where `bar`/`baz` are not guaranteed to be in a specific order.

infinisil replied:
Oh and in this case it's the order of tables themselves, yeah I didn't see anything about that in the specification. I'd still use the same reasoning though, the order of tables shouldn't have an influence.

Ma27 committed:
nixos/rust-motd: use a second attribute (`order`) for the of sections in TOML

Rather than using `priority` with `sortProperties`, a new option called
`order` defines the ordering of the sections. I.e.

    order = [ "global" "uptime" "banner" ]

means that `uptime` comes before `banner`. Please note that `global` is
for global settings and not a section. I figured that it'd be too much
magic to hide this in the implementation and ask the user to specify the
order of _each_ section in `settings` instead.

OTOH this makes the intent way clearer than priorities. Also, this
remains opt-in, the option defaults to `attrNames cfg.settings`, i.e.
all sections ordered alphabetically.

infinisil commented on nixos/modules/programs/rust-motd.nix:
```diff
@@ -5,6 +5,34 @@ with lib;
+  #   concatenated with `jq`. Now we have a JSON representation of
+  #   the config in the correct order.
+  # * this is piped to `json2toml` to get the correct format for rust-motd.
+  motdConf = pkgs.runCommand "motd.conf"
+    (orderedSections // {
+      __structuredAttrs = true;
+      nativeBuildInputs = [ pkgs.remarshal pkgs.jq ];
+    })
+    ''
+      cat "$NIX_ATTRS_JSON_FILE" \
+        | jq '${concatMapStringsSep " + " (key: ''."${key}"'') (attrNames orderedSections)}' \
+        | json2toml /dev/stdin "$out"
+    '';
```
It's generally brittle to interpolate strings, escaping can get messed up. Instead I recommend an approach like this:

```nix
  motdConf = pkgs.runCommand "motd.conf"
    {
      __structuredAttrs = true;
      inherit (cfg) settings order;
      nativeBuildInputs = [ pkgs.remarshal pkgs.jq ];
    }
    ''
      jq '.settings as $settings
          | .order
          | map({ key: "\(.)", value: $settings."\(.)" })
          | from_entries' "$NIX_ATTRS_JSON_FILE" \
        | json2toml /dev/stdin "$out"
    '';
```

No string interpolation -> no escaping problems :)

infinisil commented on nixos/modules/programs/rust-motd.nix:
```diff
@@ -27,10 +55,43 @@ in {
     settings = mkOption {
-      type = types.submodule {
+      type = types.attrsOf (types.submodule {
         freeformType = format.type;
-      };
+      });
```
I don't think this change is necessary anymore.

Ma27 replied:
We need an attribute-set at the top-level and `formts.toml.type` accepts strings and ints on the topmost level as well, so I don't think so.

infinisil replied:
Fair! Btw `freeformType` isn't needed unless there's actual options declared. So in this case `type = types.attrsOf format.type` would be enough. Probably also more efficient.

Ma27 replied:
Oof yeah, I don't even remember why I added it in the first place when writing that module. Removed.

infinisil commented on nixos/modules/programs/rust-motd.nix:
````diff
@@ -27,10 +55,43 @@ in {
+      default = attrNames cfg.settings;
+      defaultText = literalExpression "attrNames cfg.settings";
+      description = mdDoc ''
+        The order of the sections in [](#opt-programs.rust-motd.settings) implies
+        the order of sections in the motd. Since attribute sets in Nix are always
+        ordered alphabetically internally this means that
+
+        ```nix
+        {
+          uptime = { /* ... */ };
+          banner = { /* ... */ };
+        }
+        ```
+
+        will still have `banner` displayed before `uptime`.
+
+        To work around that, this option can be used to define the order of all keys,
+        i.e.
+
+        ```nix
+        {
+          order = [
+            "uptime"
+            "banner"
+          ];
+        }
+        ```
+
+        makes sure that `uptime` is placed before `banner` in the motd.
````
I guess it could be a bit more extensive actually, this feels a bit too short now, but I think you get the idea :)

Ma27 replied:
> I think this should be worded a bit more like a feature

The first question people who don't know about how Nix orders attribute-sets internally will ask is "why"? I'm pretty sure that in this case it's pretty counter-intuitive that `{ uptime = {}; banner = {}; }` doesn't result in

```toml
[uptime]
# ...
[banner]
# ...
```

if you don't know enough about how the Nix language works.

So while I agree that depending on the (apparently undefined) ordering of sections in TOML isn't very nice, I see this mostly as a workaround because of Nix's internals in this regard.

infinisil replied:
Fair, I think it would be nice for the first paragraph to be a brief explanation of what the option does, I like this:

```
The order of the sections in [](#opt-programs.rust-motd.settings).
By default they are ordered alphabetically.
```

After that could be more context.

But this is just a minor thing, I'm not gonna block on this :)

Ma27 replied:
Done!

Ma27 commented on nixos/modules/programs/rust-motd.nix:
```diff
@@ -45,14 +106,22 @@ in {
           `programs.rust-motd` is incompatible with `users.motd`!
         '';
       }
+      { assertion = length cfg.order == length (attrNames cfg.settings)
+          && all (section: cfg.settings?${section}) cfg.order;
```
When you think you know your Nix, a wild @infinisil appears :sweat_smile: 
Jokes aside, didn't even know that this was possible in Nix, TIL!

(will be part of my next push)

Ma27 committed:
nixos/rust-motd: refactor assertion and TOML generation

* `sort (<)` also works for strings (TIL!), so no need for comparing
  length and whether all keys from `cfg.settings` exist in `cfg.order`
  (slightly less overhead).

* Don't build another piece of JSON (`orderedSections`), simply use
  `cfg.settings`/`cfg.order` with `__structuredAttrs` to ensure a
  properly ordered TOML.

  This also has the upside of not having to do quote hackery.

* Also, a freeform submodule isn't strictly needed because we don't have
  any special options defined, so replacing that with
  `attrsOf format.type`.

Co-authored-by: Silvan Mosberger <github@infinisil.com>

Ma27 commented:
@infinisil I either left comments on your review comments or fixed them right away. Will also file a bug on rust-motd later :)

infinisil reviewed (approved):
Looking good!

Ma27 commented:
@Niols since you're the reporter of the original issue: does this look OK for you in its current form? If yes, I'd merge %)

Ma27 commented:
ping @Niols

Niols commented:
Hi! Sorry about this and thanks for pinging me; somehow I managed to miss this notification. The current state looks very good to me!
